### PR TITLE
FreeRTOS+TCP Adding the combined driver for SAM4E and SAME70 v2

### PR DESCRIPTION
**Pull request by htibosch**
This is a copy of # #57, just renewed it to have it properly rebased.

Description
-----------
As the title says, this PR is adding a FreeRTOS+TCP driver for 2 platforms: SAM4E and SAME70.

The driver makes use of the module ../Common/phyHandler.c, which initialises and polls the Ethernet PHY.

Related Issue
-----------
A user on the [FreeRTOS forum](https://forums.freertos.org/t/freertos-tcp-on-same70-with-asf4-atmel-start-mac-drivers/7460/16) asked for this driver. He had found a reference on SourceForge, but the ZIP file had been deleted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c [hunk 1]
```diff
@@ -0,0 +1,959 @@
...
+		xResult = 0;
+		vCheckBuffersAndQueue();
+
+		if( ( ulISREvents & EMAC_IF_ALL_EVENT ) == 0 )
+		{
+			/* No events to process now, wait for the next. */
+			ulTaskNotifyTake( pdFALSE, ulMaxBlockTime );
+		}
+
+//		if( ( ulISREvents & EMAC_IF_RX_EVENT ) != 0 )
```
[the remark this answers is not in the document]
Comment (htibosch) [reply FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c:903]: Oops, I had commented this line for a test. Now I un-commented it, and everything still works the same.

File: FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c [hunk 2]
```diff
@@ -0,0 +1,959 @@
...
+
+	ulTransmitSize = pxDescriptor->xDataLength;
+
+	pxSendPacket = (IPPacket_t *)pxDescriptor->pucEthernetBuffer;
+
+	if( ulTransmitSize > NETWORK_BUFFER_SIZE )
+	{
+		ulTransmitSize = NETWORK_BUFFER_SIZE;
+	}
+	do {
```
[the remark this answers is not in the document]
Comment (htibosch) [reply FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c:511]: I did so

File: FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c [hunk 3]
```diff
@@ -0,0 +1,959 @@
...
+ ***   FreeRTOS+TCP is functional and has been used in commercial products   ***
+ ***   for some time.  Be aware however that we are still refining its       ***
+ ***   design, the source code does not yet quite conform to the strict      ***
+ ***   coding and style standards mandated by Real Time Engineers ltd., and  ***
+ ***   the documentation and testing is not necessarily complete.            ***
+ ***                                                                         ***
+ ***   PLEASE REPORT EXPERIENCES USING THE SUPPORT RESOURCES FOUND ON THE    ***
+ ***   URL: http://www.FreeRTOS.org/contact  Active early adopters may, at   ***
+ ***   the sole discretion of Real Time Engineers Ltd., be offered versions  ***
+ ***   under a license other than that described below.                      ***
```
[the remark this answers is not in the document]
Comment (htibosch) [reply FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c:22]: Thanks for reviewing!
Hein

File: FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (AniruddhaKanhere) [reply FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/DriverSAM/NetworkInterface.c:484]: thanks

**Comment by marekr** [comment]
Well, it appears this was never actually tested on the SAM4E.
Bad define of GMAC_UR_RMIIMII
and appears the non ipconfigZERO_COPY_TX_DRIVER (SAM4E) mode wasn't tested, for example theres code that looks for p_tx_buffer which is definitely not a member of gmac_device_t

```
	switch (mode) {
		case GMAC_PHY_MII:
		case GMAC_PHY_RMII:
			p_gmac->GMAC_UR |= GMAC_UR_RMIIMII;
		break;

		default:
			p_gmac->GMAC_UR &= ~GMAC_UR_RMIIMII;
		break;
	}
```

**Comment by htibosch** [comment]
@marekr : thanks for having a look at the driver.

> Well, it appears this was never actually tested on the SAM4E.

Sure it was tested. Very recently once again, both on a SAM4E and a SAME70.

> Bad define of GMAC_UR_RMIIMII

That depends on which version of the driver that you are using.

These 2 are equivalent:

    #define GMAC_UR_RMIIMII (0x1u << 0) /**< \brief (GMAC_UR)  */
    #define GMAC_UR_MII (0x1u << 0) /**< \brief (GMAC_UR) MII Mode */

I don't know which one is the newest.

> and appears the non ipconfigZERO_COPY_TX_DRIVER (SAM4E) mode
> wasn't tested, for example theres code that looks for p_tx_buffer which
> is definitely not a member of gmac_device_t

In the [earlier version](https://github.com/FreeRTOS/FreeRTOS/tree/master/FreeRTOS-Plus/Source/FreeRTOS-Plus-TCP/portable/NetworkInterface/ATSAM4E), the non-zero-copy mode was still allowed. `p_tx_buffer` was used to hold ( a copy of ) the outgoing data.

The current driver only compiles if zero-copy is chosen:
~~~c
#if( ipconfigZERO_COPY_RX_DRIVER == 0 )
    #error This driver works optimal if ipconfigZERO_COPY_RX_DRIVER is defined as 1
#endif

#if( ipconfigZERO_COPY_TX_DRIVER == 0 )
    #error This driver works optimal if ipconfigZERO_COPY_TX_DRIVER is defined as 1
#endif
~~~
You're right, it is confusing to see code that is never compiled.

About this function in`gmac_SAM.h`:
~~~c
switch (mode) {
    case GMAC_PHY_MII:
    case GMAC_PHY_RMII:
        p_gmac->GMAC_UR |= GMAC_UR_RMIIMII;
    break;

    default:
        p_gmac->GMAC_UR &= ~GMAC_UR_RMIIMII;
    break;
}

This function was provided by Atmel as it is. It looks strange but the reason is probably: SAM4E does not support RMII. See e.g. [here](https://microchipsupport.force.com/s/article/RMII-support-on-SAM4E-devices).
And mind you, for the SAM4E: The checksum offload feature is available only for the receiver, and not for the transmitter.

**Pull request merged**